usaspending_api/etl/transaction_loaders/fpds_loader.py

```diff
@@ -208,52 +207,52 @@ def _transform_objects(broker_objects):
 def _load_transactions(load_objects):
     """returns ids for each award touched"""
     ids_of_awards_created_or_updated = set()
-    with psycopg2.connect(dsn=get_database_dsn_string()) as connection:
-        with connection.cursor(cursor_factory=psycopg2.extras.DictCursor) as cursor:
-
-            # Insert always, even if duplicative
-            # First create the records that don't have a foreign key out to anything else in one transaction per type
-            inserted_recipient_locations = bulk_insert_recipient_location(cursor, load_objects)
-            for index, elem in enumerate(inserted_recipient_locations):
-                load_objects[index]["legal_entity"]["location_id"] = inserted_recipient_locations[index]
-
-            inserted_recipients = bulk_insert_recipient(cursor, load_objects)
-            for index, elem in enumerate(inserted_recipients):
-                load_objects[index]["transaction_normalized"]["recipient_id"] = inserted_recipients[index]
-                load_objects[index]["award"]["recipient_id"] = inserted_recipients[index]
-
-            inserted_place_of_performance = bulk_insert_place_of_performance(cursor, load_objects)
-            for index, elem in enumerate(inserted_place_of_performance):
-                load_objects[index]["transaction_normalized"][
-                    "place_of_performance_id"
-                ] = inserted_place_of_performance[index]
-                load_objects[index]["award"]["place_of_performance_id"] = inserted_place_of_performance[index]
-
-            # Handle transaction-to-award relationship for each transaction to be loaded
-            for load_object in load_objects:
-
-                # AWARD GET OR CREATE
-                award_id = _matching_award(cursor, load_object)
-                if not award_id:
-                    # If there is no award, we need to create one
-                    award_id = insert_award(cursor, load_object)
-
-                load_object["transaction_normalized"]["award_id"] = award_id
-                ids_of_awards_created_or_updated.add(award_id)
-
-                # TRANSACTION UPSERT
-                transaction_id = _lookup_existing_transaction(cursor, load_object)
-                if transaction_id:
-                    # Inject the Primary Key of transaction_normalized+transaction_fpds that was found, so that the
-                    # following updates can find it to update
-                    load_object["transaction_fpds"]["transaction_id"] = transaction_id
-                    _update_fpds_transaction(cursor, load_object, transaction_id)
-                else:
-                    # If there is no transaction we create a new one.
-                    transaction_id = _insert_fpds_transaction(cursor, load_object)
-
+    connection.ensure_connection()
```

Comment:
Why is this here? Did you experience connections dropping for no apparent reason?

Reply:
@BrianZito drilling into that source code from Django, it has:
```
    def ensure_connection(self):
        """
        Guarantees that a connection to the database is established.
        """
        if self.connection is None:
            with self.wrap_database_errors:
                self.connect()
```

This is required when dropping down into the lower-level psycopg2 connection maintained and wrapped by Django in its `BaseDatabaseWrapper` class. 

Typically when using just `django.db.connection` or one of the connections from the `django.db.connections` dictionary, you're just getting the wrapper, and any calls to exposed methods on the wrapper will call this `ensure_connection()` under the hood to make sure the connection is actually connected. 

However, in this case, we're using the _unwrapped_ connection directly, so we have to do the work of _ensuring connection_ manually. 

The code that follows this line:
```
with connection.connection.cursor(cursor_factory=DictCursor) as cursor:
```
Notice the double `connection.connection`. The reason we need that one is to use the `DictCursor`, which you can't pass into the wrapped connection (only the unwrapped one), since Django doesn't overload/allow that method signature in the wrapped one.